be/tx_bulk.c

```diff
@@ -52,52 +87,158 @@ enum {
 };
 
-/* This function should be tuned using real life data. */
-static uint32_t be_tx_bulk_worker_nr(struct m0_be_tx_bulk *tb,
-                                     uint32_t              group_nr,
-                                     uint32_t              tx_per_group)
-{
-	return min_check((uint32_t)BE_TX_BULK_WORKER_MAX,
-			 tx_per_group * (group_nr + 1));
-}
+static void be_tx_bulk_finish_cb(struct m0_sm_group *grp,
+				 struct m0_sm_ast   *ast);
+static void be_tx_bulk_queue_get_cb(struct m0_sm_group *grp,
+				    struct m0_sm_ast   *ast);
+static void be_tx_bulk_init_cb(struct m0_sm_group *grp, struct m0_sm_ast *ast);
+static bool be_tx_bulk_open_cb(struct m0_clink *clink);
+static void be_tx_bulk_close_cb(struct m0_sm_group *grp, struct m0_sm_ast *ast);
+static void be_tx_bulk_gc_cb(struct m0_be_tx *tx, void *param);
+static void be_tx_bulk_queue_get_done_cb(struct m0_be_op *op, void *param);
 
 M0_INTERNAL int m0_be_tx_bulk_init(struct m0_be_tx_bulk     *tb,
                                    struct m0_be_tx_bulk_cfg *tb_cfg)
 {
-	uint32_t group_nr;
-	uint32_t tx_per_group;
-	int      rc;
+	struct be_tx_bulk_worker *worker;
+	uint64_t                  worker_partition;
+	uint64_t                  worker_locality;
+	uint64_t                  localities_nr;
+	uint32_t                  i;
+	int                       rc;
+
+	M0_PRE(M0_IS0(tb));
+	M0_PRE(tb_cfg->tbc_partitions_nr > 0);
+	/* Can't have more partitions that workers. For now. */
+	M0_PRE(tb_cfg->tbc_partitions_nr <= tb_cfg->tbc_workers_nr);
+	M0_PRE(tb_cfg->tbc_work_items_per_tx_max > 0);
+	M0_PRE(tb_cfg->tbc_q_cfg.bqc_item_length == 0);
 
 	tb->btb_cfg = *tb_cfg;
-	m0_be_domain__group_limits(tb->btb_cfg.tbc_dom,
-				   &group_nr, &tx_per_group);
-	tb->btb_worker_nr = be_tx_bulk_worker_nr(tb, group_nr, tx_per_group);
-	tb->btb_done_nr = 0;
-	tb->btb_stopping = false;
+	tb->btb_tx_open_failed = false;
 	tb->btb_done = false;
+	tb->btb_termination_in_progress = false;
 	m0_mutex_init(&tb->btb_lock);
-	M0_ALLOC_ARR(tb->btb_worker, tb->btb_worker_nr);
+	M0_ALLOC_ARR(tb->btb_worker, tb->btb_cfg.tbc_workers_nr);
 	rc = tb->btb_worker == NULL ? -ENOMEM : 0;
-	if (rc != 0)
+	if (rc != 0) {
+		m0_mutex_fini(&tb->btb_lock);
+		return M0_ERR(rc);
+	}
+	/* XXX error handling */
+	M0_ALLOC_ARR(tb->btb_q, tb->btb_cfg.tbc_partitions_nr);
+	M0_ASSERT(tb->btb_q != NULL);
+	tb->btb_cfg.tbc_q_cfg.bqc_item_length = sizeof(struct be_tx_bulk_item);
+	for (i = 0; i < tb_cfg->tbc_partitions_nr; ++i) {
+		rc = m0_be_queue_init(&tb->btb_q[i], &tb->btb_cfg.tbc_q_cfg);
+		if (rc != 0)
+			break;
+	}
+	if (rc != 0) {
+		m0_free(tb->btb_worker);
 		m0_mutex_fini(&tb->btb_lock);
+		return M0_ERR(rc);
+	}
+	m0_be_op_init(&tb->btb_kill_put_op);
+	/* XXX take it from elsewhere */
+	localities_nr = m0_fom_dom()->fd_localities_nr;
+	for (i = 0; i < tb->btb_cfg.tbc_workers_nr; ++i) {
+		worker = &tb->btb_worker[i];
+		worker_partition = i % tb->btb_cfg.tbc_partitions_nr;
```

Comment:
Is there anyway to get worker <-> partition mapping to above layer immediately after m0_be_tx_bulk_init().
Use case for this is to identify outstaring work for given partition by comparing operations added using m0_be_tx_bulk_put() Vs sum of workers completions for that partition.
currently tbc_do()/tbc_done() provide partition and worker_index parameter's but there are chances that all workers of all partitions might not have called them, when we look for outstanding work.

Reply:
There is no plan to expose the mapping because there is a plan to take work from different partitions in the same set of workers (in case of some queues are starving). So it will be everything-to-everything mapping.